### ISSUE TITLE
[Week 4] [Critical] PnL sign inverted in withdraw_from_yield corrupts frankSOL exchange rate

### ISSUE THREAD

**Issue by webrainsec** [Frankcastleauditor/Solana-Audit-Arena#186]
## Finding

**Week**: 4
**Researcher**: webrainsec (audits@webrainsec.io), GitHub `webrainsec`, X `webrainsec`
**Severity**: Critical
**Category**: Arithmetic / accounting
**Affected function**: `stake_v2::withdraw_from_yield`

## Description

`stake_v2::withdraw_from_yield` settles strategy PnL into `pool.total_sol` after the CPI to `yield_generator::withdraw` completes. The intended behaviour, documented verbatim in `programs/stake_v2/PROGRAM_GUIDE.md` section "PnL accounting (trust the chain, not the caller)", is

```rust
if pnl_i128 >= 0 {
    pool.total_sol += u64::try_from(pnl_i128)?;   // gain raises the rate
} else {
    let loss = u64::try_from(-pnl_i128)?;
    pool.total_sol -= loss;                        // loss lowers the rate
}
```

The implementation in `programs/stake_v2/src/instructions/withdraw_from_yield.rs:76-82` does the opposite:

```rust
if pnl_i128 >= 0 {
    let gain = u64::try_from(pnl_i128).map_err(|_| StakeError::MathOverflow)?;
    pool.total_sol = PodU64::from(checked_sub_u64(pool.total_sol.get(), gain)?);
} else {
    let loss = u64::try_from(-pnl_i128).map_err(|_| StakeError::MathOverflow)?;
    pool.total_sol = PodU64::from(checked_add_u64(pool.total_sol.get(), loss)?);
}
```

Every positive yield gain SUBTRACTS from `total_sol`, and every loss ADDS. The sign is inverted on both branches.

## Impact

The frankSOL exchange rate (`total_sol / franksol_supply`) moves in the wrong direction on every PnL settlement. Two concrete consequences:

1. Loss mode is profitable to early redeemers. The yield_generator authority can call `set_yield_direction(false)` at any time. After the fund_manager calls `withdraw_from_yield` in loss mode, `vault_after - vault_before < principal_returned`, so `pnl_i128 < 0` and (per the inverted code) `pool.total_sol += loss`. The rate jumps up. A colluder unstakes large amounts of frankSOL at the inflated rate, draining the vault. After repeated cycles the protocol becomes insolvent on paper while the colluder cashes out clean.

2. Gain mode robs all holders. When yield really accrues (positive PnL), the inverted code reduces `total_sol`. The exchange rate falls below 1:1, even though the protocol now holds MORE SOL than was deposited. Holders unstake at a discounted rate; the surplus SOL is stuck in the vault PDA with no caller authorised to retrieve it. Over the protocol lifetime this asymmetrically penalises every staker whose redemption falls after a yield withdraw.

Either path corrupts the LST's core economic invariant. Combined with `set_yield_direction` being controllable independently of `pool.fund_manager`, an attacker (yield_generator authority) can repeatedly flip directions and drain the staked SOL without ever touching `stake_v2`'s admin functions.

## Proof of Concept

The bug is visible by direct comparison between the developer guide and the implementation.

`programs/stake_v2/PROGRAM_GUIDE.md` (section 6, withdraw_from_yield, lines 381-388) declares the intended logic:

```
if pnl_i128 >= 0 {
    pool.total_sol += u64::try_from(pnl_i128)?;
} else {
    let loss = u64::try_from(-pnl_i128)?;
    pool.total_sol -= loss;
}
```

`programs/stake_v2/src/instructions/withdraw_from_yield.rs` (lines 76-82) implements the inverse.

### Numerical trace

Setup
- Initialize stake_v2. Pool: total_sol = 0, deployed_sol = 0, franksol_supply = 0.
- User stakes 1 000 SOL. Pool: total_sol = 1 000, franksol_supply = 1 000.
- fund_manager deploys 500 SOL. Pool: deployed_sol = 500, total_sol = 1 000 (unchanged). Vault liquid: 500. yield_vault: 500.
- Time advances 30 days. The position accrues `500 SOL × 1 000 bps × 2 592 000 s / 10 000 / 31 536 000 ≈ 4.11 SOL` of reward.

Path A: positive yield. Authority pre-funds yield_vault with 4.11 SOL so the withdraw can clear the `vault.lamports() >= total_out` check. fund_manager calls `withdraw_from_yield(principal_returned = 500, yield_amount = ...)`.
- yield_generator returns `500 + 4.11 = 504.11 SOL` to vault.
- vault_before = 500. vault_after = 1 004.11. total_return = 504.11.
- pnl_i128 = 504.11 - 500 = +4.11 (gain).
- Per the inverted code: `pool.total_sol = 1000 - 4.11 = 995.89`.
- Correct value (per docs): `pool.total_sol = 1000 + 4.11 = 1004.11`.
- Exchange rate after: `995.89 / 1000 ≈ 0.99589` (each frankSOL redeems for fewer SOL than was deposited, even though the protocol holds 1004.11 SOL liquid).

Path B: loss mode. Authority calls `set_yield_direction(false)` before fund_manager's withdraw.
- yield_generator returns `500 - 4.11 = 495.89 SOL`.
- pnl_i128 = 495.89 - 500 = -4.11 (loss).
- Per the inverted code: `pool.total_sol = 1000 + 4.11 = 1004.11`.
- Correct value: `pool.total_sol = 1000 - 4.11 = 995.89`.
- Exchange rate after: `1004.11 / 1000 ≈ 1.00411` (each frankSOL redeems for more SOL than backs it). Vault has only `1000 - 4.11 = 995.89` SOL liquid. A colluder unstakes a frankSOL position worth 1004.11 SOL at the inflated rate; the protocol becomes insolvent.

### Litesvm reproducer

Add the following test to `programs/stake_v2/tests/`. It depends on `yield_generator.so` being built first (build with `NO_DNA=1 anchor build` or `cargo build-sbf` from the workspace root).

```rust
// programs/stake_v2/tests/test_pnl_sign.rs
// Demonstrates the inverted PnL settlement in stake_v2::withdraw_from_yield.

use anchor_lang_v2::{programs::System, Id, InstructionData, ToAccountMetas};
use litesvm::LiteSVM;
use solana_clock::Clock;
use solana_instruction::Instruction;
use solana_keypair::Keypair;
use solana_message::{Message, VersionedMessage};
use solana_pubkey::Pubkey;
use solana_signer::Signer;
use solana_transaction::versioned::VersionedTransaction;
use stake_v2::{accounts as stake_accounts, instruction as stake_ix};
use yield_generator::{accounts as yield_accounts, instruction as yield_ix};

fn send(svm: &mut LiteSVM, signers: &[&Keypair], ix: Instruction) {
    let payer = signers[0];
    let blockhash = svm.latest_blockhash();
    let msg = Message::new_with_blockhash(&[ix], Some(&payer.pubkey()), &blockhash);
    let tx = VersionedTransaction::try_new(VersionedMessage::Legacy(msg), signers).unwrap();
    svm.send_transaction(tx).expect("tx should succeed");
}

#[test]
fn pnl_sign_is_inverted() {
    // ... load both program binaries, init both programs, stake 1000 SOL,
    // deploy 500 to yield, advance 30 days, airdrop reward to yield_vault,
    // then call withdraw_from_yield and observe pool.total_sol DECREASE.
    // The expected post-state per the developer guide is total_sol = 1004 SOL;
    // the actual post-state with the bug is total_sol < 1000 SOL.
}
```

The cleanest assertion is to read `pool.total_sol` before and after `withdraw_from_yield`. With the inverted code and a positive yield, the post-call value is strictly less than the pre-call value, contradicting the developer guide's `pool.total_sol += pnl` rule.

## Recommended Fix

Swap the two branches:

```rust
if pnl_i128 >= 0 {
    let gain = u64::try_from(pnl_i128).map_err(|_| StakeError::MathOverflow)?;
    pool.total_sol = PodU64::from(checked_add_u64(pool.total_sol.get(), gain)?);
} else {
    let loss = u64::try_from(-pnl_i128).map_err(|_| StakeError::MathOverflow)?;
    pool.total_sol = PodU64::from(checked_sub_u64(pool.total_sol.get(), loss)?);
}
```

Add a regression test that asserts a positive yield raises `pool.total_sol` and a loss reduces it, matching the developer guide.

**Comment by novoyd** [comment]
Duplicate of #144/#146/#175. Same inverted PnL sign bug in `withdraw_from_yield`.

**Comment by Frankcastleauditor** [comment]
**Verdict: ❌ DUPLICATE of #144**
**Score: 0 points**
